File: src/embeddedfile.go
package pdfjet

/**
 * embeddedfile.go
 *
Copyright 2020 Innovatics Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import (
	"bytes"
	"compress/zlib"
	"io"
	"io/ioutil"
	"log"
)

// EmbeddedFile is used to embed file objects in the PDF.
// The file objects must added to the PDF before drawing on the first page.
type EmbeddedFile struct {
	objNumber int
	fileName  string
	content   []byte
}

// NewEmbeddedFile is the constructor.
func NewEmbeddedFile(pdf *PDF, fileName string, reader io.Reader, compress bool) *EmbeddedFile {
	file := new(EmbeddedFile)
	file.fileName = fileName

	buf, err := ioutil.ReadAll(reader)
	if err != nil {
		log.Fatal(err)
	}

	if compress {
		var compressed bytes.Buffer
		writer := zlib.NewWriter(&compressed)
		writer.Write(buf)
		writer.Close()
		file.content = compressed.Bytes()
	} else {
		file.content = buf
	}

	pdf.newobj()
	pdf.appendString("<<\n")
	pdf.appendString("/Type /EmbeddedFile\n")
	if compress {
		pdf.appendString("/Filter /FlateDecode\n")
	}
	pdf.appendString("/Length ")
	pdf.appendInteger(len(file.content))
	pdf.appendString("\n")
	pdf.appendString(">>\n")
	pdf.appendString("stream\n")
	pdf.appendByteArray(file.content)
	pdf.appendString("\nendstream\n")
	pdf.endobj()

	pdf.newobj()
	pdf.appendString("<<\n")
	pdf.appendString("/Type /Filespec\n")
	pdf.appendString("/F (")
	pdf.appendString(fileName)
	pdf.appendString(")\n")
	pdf.appendString("/EF <</F ")
	pdf.appendInteger(pdf.getObjNumber() - 1)
	pdf.appendString(" 0 R>>\n")
	pdf.appendString(">>\n")
	pdf.endobj()

	file.objNumber = pdf.getObjNumber()

	return file
}

// GetFileName returns the file name.
func (file *EmbeddedFile) GetFileName() string {
	return file.fileName
}

File: Sources/Sources/PDFjet/EmbeddedFile.swift
/**
 *  EmbeddedFile.swift
 *
Copyright 2020 Innovatics Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import Foundation


/**
 *  Used to embed file objects.
 *  The file objects must added to the PDF before drawing on the first page.
 *
 */
public class EmbeddedFile {

    var objNumber: Int = -1
    var fileName: String?


    public init(
            _ pdf: PDF,
            _ fileName: String,
            _ stream: InputStream,
            _ compress: Bool) throws {
        self.fileName = fileName

        var baos = [UInt8]()
        var buf = [UInt8](repeating: 0, count: 4096)
        stream.open()
        while stream.hasBytesAvailable {
            let count = stream.read(&buf, maxLength: buf.count)
            if count > 0 {
                baos.append(contentsOf: buf[0..<count])
            }
        }
        stream.close()

        if compress {
            buf = baos
            baos = [UInt8]()
            _ = LZWEncode(&baos, &buf)
        }

        pdf.newobj()
        pdf.append("<<\n")
        pdf.append("/Type /EmbeddedFile\n")
        if compress {
            // pdf.append("/Filter /FlateDecode\n")
            pdf.append("/Filter /LZWDecode\n")
        }
        pdf.append("/Length ")
        pdf.append(baos.count)
        pdf.append("\n")
        pdf.append(">>\n")
        pdf.append("stream\n")
        pdf.append(baos)
        pdf.append("\nendstream\n")
        pdf.endobj()

        pdf.newobj()
        pdf.append("<<\n")
        pdf.append("/Type /Filespec\n")
        pdf.append("/F (")
        pdf.append(fileName)
        pdf.append(")\n")
        pdf.append("/EF <</F ")
        pdf.append(pdf.getObjNumber() - 1)
        pdf.append(" 0 R>>\n")
        pdf.append(">>\n")
        pdf.endobj()

        self.objNumber = pdf.getObjNumber()
    }


    public func getFileName() -> String {
        return self.fileName!
    }

}   // End of EmbeddedFile.swift
